### PR TITLE
prometheus-smartctl-exporter: init at unstable-2020-11-14

### PR DESCRIPTION
**Pull request by mweinelt**
###### Motivation for this change
Monitoring harddisk healthyness over time with Prometheus.

- Negative NixOS test, since `/dev/vda` doesn't work for <s>the exporter</s> smartmontools.
  ```json
  {
    "json_format_version": [
      1,
      0
    ],
    "smartctl": {
      "version": [
        7,
        2
      ],
      "svn_revision": "5155",
      "platform_info": "x86_64-linux-5.10.79",
      "build_info": "(local build)",
      "argv": [
        "smartctl",
        "--json",
        "--xall",
        "/dev/vda"
      ],
      "messages": [
        {
          "string": "/dev/vda: Unable to detect device type",
          "severity": "error"
        }
      ],
      "exit_status": 1
    }
  }
  ```
- Not sure if this is the right exporter to package, it has a fair share of problems and seems a bit undermaintained  :disappointed: 
	- which can be fixed by applying a few patches from open pull requests, hope they get merged soon

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/tests/prometheus-exporters.nix [hunk 1]
```diff
@@ -1015,6 +1015,25 @@ let
...
+        ];
+      };
+      exporterTest = ''
+        wait_for_unit("prometheus-smartctl-exporter.service")
+        wait_for_open_port("9633")
+        wait_until_succeeds(
+          "curl -sSf 'localhost:9633/metrics'"
+        )
+        wait_until_succeeds(
+            'journalctl -eu prometheus-smartctl-exporter.service -o cat | grep "/dev/vda: Unable to detect device type"'
```
Comment (andir) [inline nixos/tests/prometheus-exporters.nix:1032]: Isn't there a way to write a more positive test? I will no insist but there are some docs on how to simulate various SMART values for disks: https://blog.christophersmart.com/2019/12/18/kvm-guests-with-emulated-ssd-and-nvme-drives/

Can and probably pass that as additional arguments to QEMU.

Comment (mweinelt) [reply nixos/tests/prometheus-exporters.nix:1032]: The end of the blog post shows what I am seeing in QEMU VMs as well:

```
SMART support is:     Unavailable - device lacks SMART capability.
```

And so on a Hetzner VM you can compare CLI and JSON output:

```
[nix-shell:~]# smartctl --xall /dev/sda
smartctl 7.2 2020-12-30 r5155 [x86_64-linux-5.10.79] (local build)
Copyright (C) 2002-20, Bruce Allen, Christian Franke, www.smartmontools.org

=== START OF INFORMATION SECTION ===
Vendor:               QEMU
Product:              QEMU HARDDISK
Revision:             2.5+
Compliance:           SPC-3
User Capacity:        20,480,786,432 bytes [20.4 GB]
Logical block size:   512 bytes
LU is thin provisioned, LBPRZ=0
Device type:          disk
Local Time is:        Wed Nov 24 17:19:28 2021 CET
SMART support is:     Unavailable - device lacks SMART capability.
Read Cache is:        Enabled
Writeback Cache is:   Enabled

=== START OF READ SMART DATA SECTION ===
Current Drive Temperature:     0 C
Drive Trip Temperature:        0 C

Error Counter logging not supported

Device does not support Self Test logging
Device does not support Background scan results logging
```

```json
[nix-shell:~]# smartctl --json  --xall /dev/sda
{
  "json_format_version": [
    1,
    0
  ],
  "smartctl": {
    "version": [
      7,
      2
    ],
    "svn_revision": "5155",
    "platform_info": "x86_64-linux-5.10.79",
    "build_info": "(local build)",
    "argv": [
      "smartctl",
      "--json",
      "--xall",
      "/dev/sda"
    ],
    "exit_status": 4
  },
  "device": {
    "name": "/dev/sda",
    "info_name": "/dev/sda",
    "type": "scsi",
    "protocol": "SCSI"
  },
  "vendor": "QEMU",
  "product": "QEMU HARDDISK",
  "model_name": "QEMU QEMU HARDDISK",
  "revision": "2.5+",
  "scsi_version": "SPC-3",
  "user_capacity": {
    "blocks": 40001536,
    "bytes": 20480786432
  },
  "logical_block_size": 512,
  "rotation_rate": 0,
  "device_type": {
    "scsi_value": 0,
    "name": "disk"
  },
  "local_time": {
    "time_t": 1637770772,
    "asctime": "Wed Nov 24 17:19:32 2021 CET"
  },
  "temperature": {
    "current": 0,
    "drive_trip": 0
  }
}
```

**Comment by contrun** [comment]
It seems `CAP_SYS_ADMIN` and `DeviceAllow` are not enough for this exporter to run. Below is my generated systemd unit.

```
# /etc/systemd/system/prometheus-smartctl-exporter.service
[Unit]
After=network.target

[Service]
Environment="LOCALE_ARCHIVE=/nix/store/mmhbi0x5by8gih5z0sxbr2i1gm1ps39v-glibc-locales-2.33-56/lib/locale/locale-archive"
Environment="PATH=/nix/store/l6f4z8mmcnnxba8w004xn28y0vr4gdkf-coreutils-9.0/bin:/nix/store/v3lvq9hqshyldc4i6f5jy0zs0k5psbws-findutils-4.8.0/bin:/nix/store/nkwls56wcfwi1r0jnkqkvwx2zk7w3qrz-gnugrep-3.7/bin:/nix/store/vklvyr82ajbz7jm7g8dbkh62k20b0dpr-gnused-4.8/bin:/nix/store/aqwqh3drn9bf6qrs70wiyrxaawh1wcnz-systemd-249.5/bin:/nix/store/l6f4z8mmcnnxba8w004xn28y0vr4gdkf-coreutils-9.0/sbin:/nix/store/v3lvq9hqshyldc4i6f5jy0zs0k5psbws-findutils-4.8.0/sbin:/nix/store/nkwls56wcfwi1r0jnkqkvwx2zk7w3qrz-gnugrep-3.7/sbin:/nix/store/vklvyr82ajbz7jm7g8dbkh62k20b0dpr-gnused-4.8/sbin:/nix/store/aqwqh3drn9bf6qrs70wiyrxaawh1wcnz-systemd-249.5/sbin"
Environment="TZDIR=/nix/store/7wxhajvlpndhmcwrpc73sl959yb7jzbr-tzdata-2021c/share/zoneinfo"



AmbientCapabilities=CAP_SYS_ADMIN
CapabilityBoundingSet=CAP_SYS_ADMIN
DeviceAllow=/dev/nvme0
DevicePolicy=closed
DynamicUser=true
ExecStart=/nix/store/bp4061ma1h2vyz01ssrww0fz3qdb5rgq-smartctl_exporter-unstable-2020-11-14/bin/smartctl_exporter -config /nix/store/2xliikcza7ag842vrcs12rjalk3876l5-smartctl-exporter.yml

Group=smartctl-exporter
LockPersonality=true
MemoryDenyWriteExecute=true
NoNewPrivileges=true
PrivateDevices=false
PrivateTmp=true
ProcSubset=pid
ProtectClock=true
ProtectControlGroups=true
ProtectHome=true
ProtectHostname=true
ProtectKernelLogs=true
ProtectKernelModules=true
ProtectKernelTunables=true
ProtectProc=invisible
ProtectSystem=strict
RemoveIPC=true
Restart=always
RestrictAddressFamilies=AF_INET
RestrictAddressFamilies=AF_INET6
RestrictNamespaces=true
RestrictRealtime=true
RestrictSUIDSGID=true
SupplementaryGroups=disk
SystemCallArchitectures=native
SystemCallFilter=@system-service
SystemCallFilter=~@privileged @resources
UMask=0077
User=root
WorkingDirectory=/tmp
```

With `User=root`, this exporter works correctly.
```
Dez 18 22:34:01 ssg systemd[1]: Started prometheus-smartctl-exporter.service.
Dez 18 22:34:01 ssg smartctl_exporter[2445730]: [Info] Starting on 127.0.0.1:9633/metrics
```

Without `User=root`, it stops working.
```
Dez 18 22:27:41 ssg smartctl_exporter[2435759]: [Info] Starting on 127.0.0.1:9633/metrics
Dez 18 22:27:45 ssg smartctl_exporter[2435759]: [Warning] S.M.A.R.T. output reading error: exit status 2
Dez 18 22:27:45 ssg smartctl_exporter[2435759]: [Warning] The device error log contains records of errors.
Dez 18 22:27:45 ssg smartctl_exporter[2435759]: [Error] Smartctl open device: /dev/nvme0 failed: Permission denied
Dez 18 22:27:45 ssg smartctl_exporter[2435759]: [Error] smartctl returned bad data for device /dev/nvme0
Dez 18 22:28:45 ssg smartctl_exporter[2435759]: [Warning] S.M.A.R.T. output reading error: exit status 2
Dez 18 22:28:45 ssg smartctl_exporter[2435759]: [Warning] The device error log contains records of errors.
Dez 18 22:28:45 ssg smartctl_exporter[2435759]: [Error] Smartctl open device: /dev/nvme0 failed: Permission denied
Dez 18 22:28:45 ssg smartctl_exporter[2435759]: [Error] smartctl returned bad data for device /dev/nvme0
```

**Comment by mweinelt** [comment]
Use `/dev/nvme0n1` instead of `/dev/nvme0`.

**Comment by contrun** [comment]
Oops. That's it. Thank you.

**Pull request merged**